vindex/map.go

```diff
@@ -177,8 +239,54 @@ func (m *inputLogMapper) syncFromInputLog(ctx context.Context) error {
 			if err := m.walWriter.append(idx, hashes); err != nil {
 				return fmt.Errorf("failed to add index to entry for leaf %d: %v", idx, err)
 			}
+			if storeCompactRange {
+				// Periodically store the validated compact range consumed so far.
+				// The choice to align with every 256 entries is an implicit bias towards
+				// supporting tlog-tiles.
+				if err := m.walWriter.flush(); err != nil {
+					return fmt.Errorf("failed to flush the WAL: %v", err)
+				}
+				if err := m.storeState(); err != nil {
+					return fmt.Errorf("failed to store incremental state: %v", err)
+				}
+			}
 		}
 	}
+	if err := m.walWriter.flush(); err != nil {
+		return fmt.Errorf("failed to flush: %v", err)
+	}
+
+	// Calculate the root hash, and if it checks out, store the checkpoint to indicate safety.
+	hash, err := m.r.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("failed to get root hash from compact range: %v", err)
+	}
+	if !bytes.Equal(hash, cp.Hash) {
+		return fmt.Errorf("calculated hash for tree size %d is %x, but checkpoint commits to %x", cp.Size, hash, cp.Hash)
+	}
+	if err := m.db.Set([]byte(db_latestCheckpointKey), rawCp, pebble.Sync); err != nil {
+		return fmt.Errorf("failed to update state: %v", err)
+	}
+
+	klog.Infof("synced WAL to size %d", cp.Size)
+	return nil
+}
+
+func (m *inputLogMapper) storeState() error {
+	flatSlice := make([]byte, len(m.r.Hashes())*32)
```

Comment:
Use `sha256.Size`? (there are a couple of other `32`s floating around in here too, I think)

Reply:
Got em.